src/ray/rpc/grpc_client.h

```diff
@@ -40,6 +40,14 @@ namespace rpc {
     INVOKE_RPC_CALL(SERVICE, METHOD, request, callback, rpc_client); \
   }
 
+#define FIBER_RPC_CLIENT_METHOD(SERVICE, METHOD, rpc_client, SPECS)            \
```

Comment:
Will this have reduced performance compared to the RPCs run directly on gRPC threads?

How does thread pooling work for fibers, do we have a pool of fiber threads?

Reply:
In the gRPC thread, it'll move the result out, so it won't reduce the performance. 
Comparision:
- Previously, we post a job to io_thread pool. Here a job is created in asio pool.
-  Here, we pass the result to promise with move. The extra cost is **moving** the structure, which is ok. For example, this is the pub sub message, if the arena is the same, it'll just swap and we always use the default one.
```
  inline PubMessage& operator=(PubMessage&& from) noexcept {
    if (GetArena() == from.GetArena()) {
      if (this != &from) InternalSwap(&from);
    } else {
      CopyFrom(from);
    }
    return *this;
  }
```

No, we don't have a pool for fiber thread for this prototype. We just reuse the thread calling CoreWorker which can be optimized later. We should put everything not cpu intensive to one thread (fiber thread) and put cpu intensive to cpu thread pool.

Fiber's thread is like asio thread, the different part is that, for asio, they manage task, which is a closure. For fiber, it also manage context and scheduling of the tasks. In asio, we wrap everything into lambda capture, and we lose call stack. In fiber, it put things into context and makes them resumable.


In the middle of migration, we can share the same thread with asio. I mean if we go this way we can do it granularly.